rfcs/0096-experimental-nix-derivations.md

```diff
@@ -0,0 +1,81 @@
+---
+feature: experimental-nix-derivations
+start-date: 2021-06-25
+author: gytis-ivaskevicius
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+---
+
+# Summary
+[summary]: #summary
+
+Add a way for users to use experimental Nix features with improved user-experience
+and better communication to the user that they are using an experimental feature.
+
+# Motivation
+[motivation]: #motivation
+
+Motivation:
+- Revert of (this)[https://github.com/NixOS/nixpkgs/pull/123898].
+- Noticed quite a few peeps who were very supportive of [this](https://github.com/NixOS/nixpkgs/pull/120141) PR (which got reverted).
+- Lack of visibility for the end-users that they are using experimental features.
+
+Expected outcome:
+- Improved UX.
+- Reduced risk of users using experimental features in production environments.
+- Reduced non-bug questions/issues on common support channels.
+
+# Detailed design
+[design]: #detailed-design
+
+`nixUnstable` derivation definitions should be updated in such a way so end users would be able to overwrite features with something along the lines of this:
+```nix
+nixUnstable.override { experimentalFeatures = ["flakes"]; }
+```
+
+What's so special about this override:
+- Makes specified experimental features enabled by default. (A little similar to [this](https://github.com/NixOS/nixpkgs/pull/120141) PR)
+- When using these overwritten derivations warning should be shown.
+
+
+Warning expected to look something like this:
+```
+WARNING: You are using experimental Nix features which are up to a subject to change. For more information visit https://nixos.org/xyz
+```
+
+A webpage is expected to contain information as such:
+- Support channels if any.
+- A little rundown of the current state of the feature(s).
+- Some "URL dump" kind of wiki page that that would contain a list of links to issues with existing bugs/workarounds/edge-cases.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+If we take Nix flakes as an example:
+The additional line to `nix.conf` is not a deal-breaker for anyone who decides to
+use Nix Flakes. Even if someone were to do so in a production environment is
+likely to get by unnoticed. On another hand, if there is an explicit warning it is
+likely to make multiple peeps question their decisions.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+I can't think of any drawbacks.
```

Comment:
Warnings may be more annoying than config to some people https://github.com/NixOS/nixpkgs/pull/120141#issuecomment-845942019

Reply:
I also don't like having a warning there every time. I don't know how warnings are implemented exactly but it could lead to overlooking other warnings depending on how you check for them (maybe even automatically?)

Reply:
I think a simple switch in the nix.conf to supress the warning would be fairly simple no?

Reply:
It's just a one-line, and I guess something similar to `warn-dirty` option could be implemented

Reply:
we could also make `warn-dirty` more general; say `supress-warnings`

Reply:
Can we make it right for everyone w.r.t. warnings? Probably we can, but in my opinion it might be that we also end up deciding to put some people off worse, here.

One of the motivating items is:

> - Reduced risk of users using experimental features in production environments

As I explained in https://github.com/NixOS/rfcs/pull/96#discussion_r662441595, the config currently is pretty much contextual (that is: "impure") state. And people might indeed forget about their current setup, especially if they switch a lot. A warning is an appropriate solution to that problem, even if it annoys others.

Just mentioning that it _might_ be a possibility.